### PR TITLE
MueLu/regionMG: separate setup and apply of coarse solver

### LINKED ISSUE #5072: MueLu: separate setup and solve of regionMG coarse level solver
## Enhancement
@trilinos/muelu 

On the coarse level, we need to perform a direct solve on the composite system. Right now, this is done by creating a Amesos2 KLU solver in every visit to the coarsest level.

Steps to do:
- Move the setup phase of the Amesos2 KLU solver to the setup of the hierarchy and store the factorizations
- Only leave the apply phase of the direct solver in the V-Cycle

Overall, this aims at reducing unnecessary (and duplicate) cost.

### PR DESCRIPTION
**Pull request by mayrmt**
@trilinos/muelu 

## Description

Split setup and solve of the composite coarse level solver into two parts. Do the setup during hierarchy construction and just apply the solver during the V-cycle.

I also have removed commented and obsolete code from my previous unsuccessful attempts to create a coarse solver.

## Motivation and Context

The matrix does not change, it's sufficient to create the solver once and then just re-use multiple times. This should save unecessary setup costs.

## Related Issues

* Closes #5072 

## How Has This Been Tested?

Build and regionMG tests pass locally.

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp [hunk 1]
```diff
@@ -1326,6 +1383,48 @@ void createRegionHierarchy(const int maxRegPerProc,
 } // createRegionHierarchy
 
 
+// Wrapper to be used from the Matlab-based driver
```
Comment (lucbv) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: That is not a bad idea, as you realize, I was using default values for some parameters to comply with the matlab driver but this looks better.

Comment (mayrmt) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: I had trouble to add another argument with default value. It always becomes difficult if there are multiple default values...

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp [hunk 2]
```diff
@@ -1314,6 +1367,10 @@ void createRegionHierarchy(const int maxRegPerProc,
                               regMatrices,
                               coarseCompOp);
 
+  std::cout << mapComp->getComm()->getRank() << " | MakeCoarseCompositeDirectSolver ..." << std::endl;
+
+  coarseSolver = MakeCompositeDirectSolver(coarseCompOp);
```
Comment (lucbv) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: I think we should go a step further and actually do the LU factorization here, calling the following methods:
```
coarseSolver->symbolicFactorization();
coarseSolver->numericFactorization();
```
You can see an example in Amesos2 [here](https://trilinos.org/docs/r12.12/packages/amesos2/doc/html/TwoPartSolve_8cpp-example.html).

Comment (mayrmt) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: I call exactly these two methods inside `MakeCompositeDirectSolver()` as the last step before returning the solver object.

Comment (lucbv) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: Yes I see that now! Great then it's good with me : )

**Pull request merged**